packages/app/src/background/services/injector/Injector.ts

```diff
@@ -25,49 +47,72 @@ export class InjectorService {
+          urlOrigin,
+        });
+        await this.approvalService.add({ urlOrigin: urlOrigin!, canSkipApprove: skipApprove });
+      } else {
+        await this.approvalService.add({ urlOrigin: urlOrigin!, canSkipApprove });
+      }
 
-    return payload;
+      if (!isConnected || isChangeIdentity) {
+        await this.connectionService.connectRequest({}, { urlOrigin: urlOrigin! });
+      }
+    } catch (error) {
+      throw new Error(`CryptKeeper: error in the connect request, ${(error as Error).message}`);
+    }
   };
 
-  getConnectedIdentityMetadata = async (
-    _: unknown,
-    { urlOrigin }: IZkMetadata,
-  ): Promise<ConnectedIdentityMetadata | undefined> => {
-    const { isApproved } = this.injectorHandler.getConnectionApprovalData({ urlOrigin });
+  getConnectedIdentityMetadata = (_: unknown, { urlOrigin }: IZkMetadata): ConnectedIdentityMetadata | undefined => {
+    const { isApproved, isConnected } = this.getConnectionData({ urlOrigin });
 
-    if (!isApproved) {
+    if (!isApproved || !isConnected) {
       return undefined;
     }
 
-    return this.injectorHandler.getConnectedIdentityMetadata({}, { urlOrigin });
+    const connectedIdentity = this.connectionService.getConnectedIdentity(urlOrigin!);
+    return pick(connectedIdentity?.metadata, ["name"]) as ConnectedIdentityMetadata | undefined;
   };
 
-  connect = async ({ isChangeIdentity }: IConnectionOptions, { urlOrigin }: IZkMetadata): Promise<void> => {
-    const { isApproved, canSkipApprove } = await this.injectorHandler.checkApproval({ urlOrigin });
+  private newRequest = async <T>(type: PendingRequestType, payload: unknown): Promise<T> => {
+    const response = await this.requestManager.newRequest(type, payload);
+    await this.browserService.closePopup();
 
-    try {
-      if (isApproved) {
-        await this.injectorHandler.getApprovalService().add({ urlOrigin: urlOrigin!, canSkipApprove });
-      } else {
-        const hostPermission = (await this.injectorHandler.newRequest(PendingRequestType.APPROVE, {
-          urlOrigin,
-        })) as IHostPermission;
+    return response as T;
+  };
 
-        await this.injectorHandler.getApprovalService().add({
-          urlOrigin: hostPermission.urlOrigin,
-          canSkipApprove: hostPermission.canSkipApprove,
-        });
-      }
+  private getConnectionData = ({ urlOrigin }: IZkMetadata): IConnectionData => {
+    if (!urlOrigin) {
+      throw new Error("CryptKeeper: Origin is not set");
+    }
+
+    const isApproved = this.approvalService.isApproved(urlOrigin);
+    const canSkipApprove = this.approvalService.canSkipApprove(urlOrigin);
+    const identity = this.connectionService.getConnectedIdentity(urlOrigin);
+
+    return { isApproved, isConnected: Boolean(identity), canSkipApprove };
+  };
 
-      const connectedIdentity = await this.injectorHandler.getZkIdentityService().getConnectedIdentity();
+  private checkUnlockStatus = async () => {
```

Comment:
I think we can rename this one to be just `ensure` and it would be the injector service middleware.

```suggestion
// INJECTOR
  ensure = async (payload: unknown): unknown => {
      const { isUnlocked, isInitialized } = await this.lockerService.getStatus();

    if (!isUnlocked) {
      await this.browserService.openPopup();
      await this.lockerService.awaitUnlock();

      if (isInitialized) {
        await this.approvalService.awaitUnlock();
        await this.zkIdentityService.awaitUnlock();
        await this.connectionService.awaitUnlock();
        await this.browserService.closePopup();
      }
    }
    
    return payload;
  }
```

```ts
    this.handler.add(
    RPCExternalAction.CONNECT,  
    this.injectorService.ensure,  
    this.injectorService.connect);
```

And then we can remove this ` await this.checkUnlockStatus();` from connect function in the injector service.

Reply:
This logic isn't shared for other services. It stays the same as it was before, nothing was changed here.